mitiq/cdr/data_regression.py

```diff
@@ -0,0 +1,294 @@
+    return (X_data, Y_data)
+
+
+# TODO: again this function below is not really necessary, but is quite useful
+#  and follows the structure of the above format.
+
+
+def construct_circuit_data_floats(
+    circuit_data: List[Dict], observable: np.ndarray, noise_levels: int = 1,
+) -> np.ndarray:
+    """Returns training data now as two arrays of floats to be used in the
+    regression (raw_training_data, simulated_training_data)
+    Args:
+        circuit_data: list of dictionary of counts for circuit of interest
+                      and all noise levels.
+        observable: option to be passed to use defined observable function that
+                    defines how to calculate the value of an observable from
+                    the counts.
+        noise_levels: number of noise levels to be used in the regression.
+    """
+    circuit_data_floats = []
+    for result in circuit_data:
+        obs_raw = calculate_observable(result, observable)
+        circuit_data_floats.append(obs_raw)
+    return circuit_data_floats
+
+
+def find_optimal_parameters(
+    X_data: np.ndarray, y_data: np.ndarray, intercept: bool = True
+):
+    """Returns the optimal parameters calculated by computing a regression
+    on the training data.
+    Args:
+        X_data: array of noisy observables at various noise levels.
+        y_data: array of exact observables.
+        intercept: whether or not to include and intercept in the fitting
+                   function.
+    """
+    noise_levels = len(X_data[0][:])
+    X_data = X_data.T
+    if intercept:
+        params_0 = np.zeros((noise_levels + 1))
+    else:
+        params_0 = np.zeros((noise_levels))
+
+    # Here need to use some package that does the regression.
+    popt, pcov = curve_fit(
+        lambda x, *params: wrapper_fit_func(
+            x, noise_levels, params, intercept
+        ),
+        X_data,
+        y_data,
+        p0=params_0,
+    )
+    return popt
+
+
+def f_intercept(X_data: np.ndarray, params: List) -> float:
+    """ Function used to map noisy to exact expectation values. Fitted using
+    data from the training circuits.
+    Args:
+        X_data: array of noisy observables are various noise levels.
+        *params: parameters of function.
+    """
+    return sum(a * x for a, x in zip(params, X_data)) + params[-1]
+
+
+def f(X_data: np.ndarray, params: np.ndarray) -> float:
+    """ Function used to map noisy to exact expectation values. Fitted using
+    data from the training circuits. No intercept.
+    Args:
+        X_data: array of noisy observables are various noise levels.
+        *params: parameters of function.
+    """
+    return sum(a * x for a, x in zip(params, X_data))
+
+
+def wrapper_fit_func(x, noise_levels, *args, intercept: bool = True):
+    if intercept:
+        a = list(args[0][: noise_levels + 1])
+        return f_intercept(x, a)
+    else:
+        a = list(args[0][:noise_levels])
+        return f(x, a)
```

Comment:
Can't you just have one function which always has an intercept that could be zero?

Reply:
Do you mean setting the intercept to zero? Or allowing it to fit and expecting it to be zero? 

Having the extra freedom to fit an intercept changes the fit. We found for vnCDR including a constant makes the ansatz less natural to motivate. Whereas for CDR it is more natural. In experiments on real data I have normally found including a constant to be better for both vnCDR and CDR. 

Personally I am not tied keeping this option if you think it is messy and arbitrary. In the future it would be ideal for a user to define their own functions to use for the regression anyway.

Reply:
Gotcha. I'm fine leaving the two functions then. 

> In the future it would be ideal for a user to define their own functions to use for the regression anyway.

If this is the case, the function should really be an argument to `execute_with_cdr`. I suggest:

1. Add `regression: Callable[..., float] = linear_fit` to `execute_with_cdr`. (`linear_fit` is currently called `f`).
2. Do the [fitting with scipy](https://github.com/unitaryfund/mitiq/pull/677/files#diff-30fe2c40c2db4727e319f1930148eff86562fab3dde123ff71ebee6d97ad4b14R228-R235) inside `execute_with_cdr`, or keep the `find_optimal_paramaters` helper function but generalize it to any regression (right now it's specific to `wrapper_fit_function`).
3. Remove `wrapper_fit_function` - after (1) and (2) you don't need it anymore.

What do you think? Does this sound viable?

Reply:
I think this is a very good suggestion.

Reply:
After thinking about this a little more I think it might be tricky to generalise the fitting process. It might be better to have the user write the regression functions and how to calculate the optimal parameters. I propose combining the `find_optimal_parameters` function and the `linear_fit` function. I think we can still call this `linear_fit` but just involve the fitting with scipy inside that function.

Reply:
actually, just realised this won't work because we also need access to the function form to complete the mitigation. So we would need two user defined functions. The functional form to fit and the function to calculate the optimal parameters of this function from the training data.